File: nowcoder/78.反转链表.py
# -*- coding:utf-8 -*-
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    # 返回ListNode
    def ReverseList(self, pHead):
        new_head = ListNode(0)

        while pHead is not None:
            a = pHead
            pHead = pHead.next

            a.next = new_head.next
            new_head.next = a

        return new_head.next

File: nowcoder/78.反转链表.cpp
#include <iostream>

struct ListNode {
    int val;
    struct ListNode* next;
    ListNode(int x)
        : val(x)
        , next(NULL)
    {
    }
};

class Solution {
public:
    ListNode* ReverseList(ListNode* pHead)
    {
        if (pHead == NULL) {
            return NULL;
        }

        ListNode* pre = NULL;
        ListNode* p = pHead;

        while (p) {
            ListNode* next = p->next;
            p->next = pre;

            pre = p;
            p = next;
        }
        return pre;
    }
};